backends/dpdk/dpdkProgram.cpp

```diff
@@ -0,0 +1,282 @@
+                                ingress_deparser_converter->getInstructions(),
+                                egress_parser_converter->getInstructions(),
+                                egress_converter->getInstructions(),
+                                egress_deparser_converter->getInstructions()});
+  statements.push_back(s);
+
+  return new IR::DpdkAsmProgram(
+      headerType, structType, ingress_converter->getActions(),
+      ingress_converter->getTables(), statements, collector->get_globals());
+}
+
+const IR::Node *ConvertToDpdkProgram::preorder(IR::P4Program *prog) {
+  // std::cout << prog << std::endl;
+  dpdk_program = create(prog);
+  return prog;
+}
+
+bool ConvertToDpdkParser::preorder(const IR::P4Parser *p) {
+  for (auto l : p->parserLocals) {
+    collector->push_variable(new IR::DpdkDeclaration(l));
+  }
+  std::unordered_map<cstring, int> degree_map;
+  std::unordered_map<cstring, IR::ParserState> state_map;
+  std::vector<IR::ParserState> stack;
+  for (auto state : p->states) {
+    if (state->name == "start")
+      stack.push_back(*state);
+    degree_map.insert({state->name.toString(), 0});
+    state_map.insert({state->name.toString(), *state});
+  }
+  for (auto state : p->states) {
+    if (state->selectExpression) {
+      if (state->selectExpression->is<IR::SelectExpression>()) {
+        auto select = state->selectExpression->to<IR::SelectExpression>();
+        for (auto pair : select->selectCases) {
+          auto got = degree_map.find(pair->state->path->name);
+          if (got != degree_map.end()) {
+            got->second++;
+          }
+        }
+      } else if (auto path =
+                     state->selectExpression->to<IR::PathExpression>()) {
+        auto got = degree_map.find(path->path->name);
+        if (got != degree_map.end()) {
+          got->second++;
+        }
+      }
+    }
+  }
+  degree_map.erase("start");
+  state_map.erase("start");
+
+  while (stack.size() > 0) {
+    auto state = stack.back();
+    stack.pop_back();
+
+    // the main body
+    auto i = new IR::DpdkLabelStatement("L_" + state.name.toString());
+    add_instr(i);
+    auto c = state.components;
+    for (auto stat : c) {
+      DPDK::ConvertStatementToDpdk h(refmap, typemap, 0, this->collector,
+                                     csum_map);
+      stat->apply(h);
+      for (auto i : h.get_instr())
+        add_instr(i);
+    }
+    if (state.selectExpression) {
+      if (auto e = state.selectExpression->to<IR::SelectExpression>()) {
+        const IR::Expression *switch_var;
+        switch_var = e->select->components[0];
+        for (auto v : e->selectCases) {
+          if (!v->keyset->is<IR::DefaultExpression>()) {
+            add_instr(new IR::DpdkJmpEqualStatement("L_" + v->state->path->name,
+                                                    v->keyset, switch_var));
+          } else {
+            auto i = new IR::DpdkJmpStatement("L_" + v->state->path->name);
```

Comment:
You are not looking at the select expression at all here. where is this done?

Reply:
@gordonwucn please take a look.

Reply:
`switch_var` is the SelectExpression. This is the statement that compares if SelectExpression equals one of the keys. `new IR::DpdkJmpEqualStatement("L_" + v->state->path->name, v->keyset, switch_var)`

Reply:
But you are only looking at component[0]. You may want to give an error if there are more components?

Reply:
Thanks for mentioning! @hanw Could you throw an error when `e->select->components` is longer than 1?

Reply:
added.